lib/guardian/token/jwt.ex

```diff
@@ -124,6 +146,66 @@ defmodule Guardian.Token.Jwt do
   @type_key "typ"
   @default_ttl {4, :weeks}
 
+  defmodule SecretFetcher do
+    @moduledoc """
+    Provides a behaviour that specifies how to fetch the secret for the token
+
+    `use Guardian.Token.JWT.SecretFetcher` to provide default implementations of each function
+    """
+
+    @doc """
+    fetch_siginig_secret fetches the secret to sign.
+    """
+    @callback fetch_signing_secret(module, opts :: Guardian.options()) ::
+                {:ok, term} | {:error, :secret_not_found}
+
+    @doc """
+    fetch_verifying_secret fetches the secret to verify a token.
+    It is provided with the tokens headers in order to lookup the secret.
+    """
+    @callback fetch_verifying_secret(module, token_headers :: map, opts :: Guardian.options()) ::
+                {:ok, term} | {:error, :secret_not_found}
+
+    defmacro __using__(_opts \\ []) do
+      quote do
+        alias Guardian.Token.Jwt.SecretFetcher.SecretFetcherDefaultImpl, as: DI
+
+        def fetch_signing_secret(mod, opts),
+          do: DI.fetch_signing_secret(mod, opts)
+
+        def fetch_verifying_secret(mod, token_headers, opts),
+          do: DI.fetch_verifying_secret(mod, token_headers, opts)
+
+        defoverridable fetch_signing_secret: 2, fetch_verifying_secret: 3
+      end
+    end
+  end
+
+  defmodule SecretFetcher.SecretFetcherDefaultImpl do
+    @moduledoc false
+    use Guardian.Token.Jwt.SecretFetcher
```

Comment:
I think it's possible to get rid of the macro altogether, but please excuse me if this makes no sense as I've just started learning Elixir and using Guardian to verify Firebase tokens...

Isn't it enough with `@behaviour Guardian.Token.Jwt.SecretFetcher`? After all `fetch_secret_fetcher` is already referencing `SecretFetcherDefaultImpl` by default. This way we can get rid of the macro, which also resembles OO inheritance.